src/Microsoft.Azure.SignalR.Emulator/HubEmulator/HubProxyHandler.cs

```diff
@@ -0,0 +1,399 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.AspNetCore.SignalR.Protocol;
+using Microsoft.Azure.SignalR.Serverless.Common;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.SignalR.Emulator.HubEmulator
+{
+    internal class HubProxyHandler<THub> : HubConnectionHandler<THub> where THub: Hub
+    {
+        internal static TimeSpan DefaultHandshakeTimeout => TimeSpan.FromSeconds(15);
+        internal static TimeSpan DefaultKeepAliveInterval => TimeSpan.FromSeconds(15);
+        internal static TimeSpan DefaultClientTimeoutInterval => TimeSpan.FromSeconds(30);
+
+        internal const int DefaultMaximumMessageSize = 32 * 1024;
+        internal const int DefaultStreamBufferCapacity = 10;
+        private readonly IOptionsMonitor<UpstreamOptions> _upstreamSettings;
+        private readonly HubLifetimeManager<THub> _lifetimeManager;
+        private readonly ILoggerFactory _loggerFactory;
+        private readonly ILogger _logger;
+        private readonly IHubProtocolResolver _protocolResolver;
+        private readonly HubOptions<THub> _hubOptions;
+        private readonly HubOptions _globalHubOptions;
+        private readonly IUserIdProvider _userIdProvider;
+        private readonly bool _enableDetailedErrors;
+        private readonly HttpServerlessMessageHandler<THub> _upstream;
+
+        private readonly string _hubName = typeof(THub).Name;
+
+        private readonly List<string> _defaultProtocols = new List<string>();
+
+        public HubProxyHandler(
+            IOptionsMonitor<UpstreamOptions> upstreamSettings,
+            HubLifetimeManager<THub> lifetimeManager,
+            IHubProtocolResolver protocolResolver,
+            IOptions<HubOptions> globalHubOptions,
+            IOptions<HubOptions<THub>> hubOptions,
+            ILoggerFactory loggerFactory,
+            IUserIdProvider userIdProvider,
+            IServiceScopeFactory serviceScopeFactory,
+            HttpServerlessMessageHandler<THub> upstream)
+            : base(lifetimeManager, protocolResolver, globalHubOptions, hubOptions, loggerFactory, userIdProvider, serviceScopeFactory)
+        {
+            _protocolResolver = protocolResolver;
```

Comment:
protocolResolver ?? throw XXXException(...) ?

Reply:
This file is actually a modified version of https://github.com/dotnet/aspnetcore/blob/v3.1.7/src/SignalR/server/Core/src/HubConnectionHandler.cs, so keep it the similar style as the SignalR one. Add a comment in the class to explain it.